sdks/python/apache_beam/runners/pipeline_context.py

```diff
@@ -233,12 +233,20 @@ def requirements(self):
   # rather than an actual coder. The element type is required for some runners,
   # as well as performing a round-trip through protos.
   # TODO(BEAM-2717): Remove once this is no longer needed.
-  def coder_id_from_element_type(self, element_type):
-    # type: (Any) -> str
+  def coder_id_from_element_type(
+      self, element_type, requires_deterministic_key_coder=None):
+    # type: (Any, Optional[str]) -> str
     if self.use_fake_coders:
       return pickler.dumps(element_type).decode('ascii')
     else:
-      return self.coders.get_id(coders.registry.get_coder(element_type))
+      coder = coders.registry.get_coder(element_type)
+      if requires_deterministic_key_coder:
+        coder = coders.TupleCoder([
```

Comment:
I don't understand this part. Why is it safe to assume that any coder here will have a key and value coder?

Reply:
If it requires a deterministic key, it must be a coder with a key (and value).

Reply:
Got it, requires_deterministic_key_coder is set by the ForceKvInputTypes visitor.